### DISCUSSION TITLE
MXAPI: Add functions to create/rename/delete Mailboxes

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#4007] [Development]
Create three new API functions:
- `mbox_create()`
- `mbox_rename()`
- `mbox_delete()`

For each function:
- Decide which Mailbox types are supported
  (compressed, imap, maildir, mbox, mh, mmdf, nntp, notmuch, pop)
- Decide what parameters are needed

Implement the backend functions
Update the browser to use them

**Comment by ngortheone** [comment]
Hey @flatcap. I want to try myself in C development(I have mostly golang/java/scala experience). Where would you recommend starting diving into the codebase with this issue in mind?

**Comment by flatcap** [comment]
Hi @ngortheone.  Welcome to NeoMutt!

> I want to try myself in C development

great

> Where would you recommend starting

hmm... Most of the items on my 'to do' list are either trivial or mind-bending.
(I don't want to scare you off before you even start :-)

I'll pick out some good starter tasks and get back to you.

If you use IRC, I recommend joining us in `#neomutt` on `irc.freenode.net`

Cheers, flatcap.

**Comment by flatcap** [comment]
Here are some unsorted notes I made, ages ago:

------------------------------------------------------------

`int mbox_create(account, path, name, flags)`
- account: to create Mailbox on
- path:    location (optional)
- name:    of new Mailbox
- flags:   ?

Support:

- Account type
  - imap
- `$mbox_type`
  - maildir/mh
  - mbox/mmdf
- Not implemented
  - compressed
  - nntp
  - notmuch
  - pop

Notifications:
    MAILBOX_NEW on Mailbox -> Account

------------------------------------------------------------

`int mbox_rename(mailbox, name, flags)`
- mailbox: to rename
- name:    new name
- flags:   ?

Support:

- Account type
  - imap
  - maildir/mh
  - mbox/mmdf
  - compressed
- Not implemented
  - nntp
  - notmuch
  - pop

Notifications:
    MAILBOX_RENAME on Mailbox -> Account

------------------------------------------------------------

`int mbox_delete(mailbox)`
- mailbox: to delete

Support:

- Account type
  - imap
  - maildir/mh
  - mbox/mmdf
  - compressed
- Not implemented
  - nntp
  - notmuch
  - pop

Notifications:
    MAILBOX_DELETE on Mailbox -> Account

------------------------------------------------------------

can the browser create nested mailboxes?
    possibly
    the user enters a name with path delimiters
    the backend needs to figure out if that's valid or not

when in a folder A, viewing children X, Y, Z
    where does new mailbox N get created?
    how to get it created in X without opening X?
        descend?

frontend asks for name, not backend

future need api completion for
	COMP_CREATE
	COMP_DELETE
	etc

**Comment by matthewhughes934** [comment]
update for `mbox_create`, following me reading around a bit + #2853, I'd be happy to hear any thoughts/objections.

```c

/**
 * enum MxCreateReturns - Return values for mbox_create()
 */
enum MxCreateReturns
{
  MX_CREATE_OK,           ///< Create succeeded
  MX_CREATE_EXISTS,       ///< Create failed because a matching mailbox already exists
  MX_CREATE_BAD_NAME,     ///< Create failed because the name was invalid, e.g. ''
  MX_CREATE_SYS_ERROR,    ///< Create failed because of a system error, errno set
  MX_CREATE_ERROR,        ///< Create failed because of a general error
};


  /**
   * mbox_create - create a new mailbox
   * @param[in] a Account to create the mailbox on
   * @param[in] m_parent parent Mailbox to create under
   * @param[in] name The name of the mailbox
   * @param[out] new_ptr pointer to the mailbox to be created
  */
  enum MxCreateReturns (*mbox_create)       (struct Account *a, struct Mailbox *m_parent, const char *name, struct Mailbox **new_ptr);
```

## `maildir`

Create a new dir under `mailbox_path(m_parent)` with subdirs `cur`, `tmp`, `new`

## `mh`

Create a new dir under `mailbox_path(m_parent)` with `.mh_sequences`

## `mbox/mmdf`

Haven't thought _too_ much about these. Maybe just create files in `dirname(mailbox_path(m_parent))`? Then I'm _assuming_ it's fine to leave them to wait for their first message, since since to satisfy the corresponding [`MxOps::path_probe`](https://github.com/neomutt/neomutt/blob/master/mbox/mbox.c#L1619) the files should start with their corresponding separator.

## `IMAP`

Maybe just parent path + delim + name?

**Comment by flatcap** [comment]
> mbox/mmdf

An empty file is a valid mbox/mmdf mailbox.
How it's interpreted, when you come to save to it, will depend on your `$mbox_type` value

> IMAP

I don't think we want to worry about delimiters, just a parent (or NULL) and a name.
i.e. We won't allow users to try to create deep folders without the intermediate folders first.
